.github/workflows/bats.yaml

```diff
@@ -50,42 +96,167 @@ jobs:
+        DATA="$(gpg --batch --with-colons --with-keygrip --list-secret-keys)"
+        FINGERPRINT="$(awk -F: '/^fpr:/ { print $10 ; exit }' <<< "${DATA}")" # spellcheck-ignore-line
+        GRIP="$(awk -F: '/^grp:/ { print $10 ; exit }' <<< "${DATA}")"
+
+        # Save the password
+        gpg-connect-agent --verbose "PRESET_PASSPHRASE ${GRIP} -1 00" /bye
+
+        # Initialize pass
+        pass init "${FINGERPRINT}"
+
+    - name: Normalize test name
+      run: |
+        t="${{ matrix.name }}"
+        if [[ ! -r "tests/$t" ]] && [[ -r "tests/${t}.bats" ]]; then
+          t="${t}.bats"
+        fi
+        echo "RD_TEST=$t" >> "$GITHUB_ENV"
+      working-directory: bats
+
+    - name: "macOS: Set startup command"
+      if: runner.os == 'macOS'
+      run: echo "BATS_COMMAND=$BATS_COMMAND" >> "$GITHUB_ENV"
+      env:
+        BATS_COMMAND: exec
+    - name: "Linux: Set startup command"
+      if: runner.os == 'Linux'
+      run: echo "BATS_COMMAND=$BATS_COMMAND" >> "$GITHUB_ENV"
+      env:
+        BATS_COMMAND: >-
+          exec xvfb-run --auto-servernum
+          --server-args='-screen 0 1280x960x24'
+
+    - name: Run BATS
+      run: >-
+        ${{ env.BATS_COMMAND }}
+        ./bats-core/bin/bats
+        --gather-test-outputs-in "${LOGS_DIR}"
+        --formatter tap
+        --report-formatter tap
+        'tests/${{ env.RD_TEST }}'
       env:
         RD_CONTAINER_ENGINE:     ${{ matrix.engine }}
         RD_CAPTURE_LOGS:         "true"
         RD_TAKE_SCREENSHOTS:     "true"
-        RD_USE_IMAGE_ALLOW_LIST: "true"
+        RD_TRACE:                "true"
+        RD_USE_GHCR_IMAGES:      "true"
+        RD_USE_RAMDISK:          "true"
+        LOGS_DIR:                ${{ github.workspace }}/logs
+        GITHUB_TOKEN:            ${{ github.token }}
+        BATS_COMMAND:            ${{ env.BATS_COMMAND }}
+        RD_TEST:                 ${{ env.RD_TEST }}
+      working-directory: bats
+      timeout-minutes: 120
 
     - name: Consolidate logs
       if: ${{ !cancelled() }}
       run: |
         # bats/logs may not exist if the workflow is being tested with e.g. tests/helpers/utils.bats
-        if [ -d "$TMPDIR/bats/logs" ]; then
-            cp -R "$TMPDIR/bats/logs/" logs
+        if [ -d "bats/logs" ]; then
+            cp -R "bats/logs/" logs
         fi
-        cp "$TMPDIR/bats/report.tap" logs
-        "$TMPDIR/bats/scripts/rename-logs.sh" logs
+        cp "bats/report.tap" logs
+        "bats/scripts/rename-logs.sh" logs
+        # Strip any `.bats` suffix if we're running individual test files.
+        echo "\`${NAME}\` (${OS} ${ENGINE})" > logs/test-name.txt
+        mv version.txt logs/
+      env:
+        NAME: ${{ matrix.name }}
+        OS: ${{ matrix.host }}
+        ENGINE: ${{ matrix.engine }}
+
+    - name: Calculate log name
+      id: log_name
+      if: ${{ !cancelled() }}
+      run: |
+        for ch in '"' ':' '/' '<' '>' '|' '\*' '\?'; do
+          name="${name//${ch}/_}"
+        done
```

Comment:
This looks very similar to what `bats/scripts/rename-logs.sh` is doing; can this be combined into a single script? Just to make sure we deal with the same set of characters in both places.

Reply:
I'll see if I can make `rename-logs.sh` a bit more flexible (to cover this use case).